Add pattern matching in the filtering plugin

dcaliste opened the pull request:
@pvuorela , this is done to continue the discussion started elsewhere on the pattern matching.

The first commit introduces a minimal (but enough for my needs) pattern matching capability to the filtering plugin. It is not handling how these patterns are added or removed from the DConf entries. As you see, you can still list the DConf values, and display in UI only strings starting with '+' or a digit, or don't display strings starting with '^' or being '*'. The priority is always given to exact matching, so an UI adding and removing exact matching entries will continue to behave exactly like before.

The second commit, besides introducing a test for these filtering combinations, is more about defining an API that can internally handle the DConf entries, while being imperative : like ignore this number (if it was whitelisted, it is removed…), reject these numbers starting with this prefix… To unblock a number, I introduce a subtle difference between "stop blocking this" and "actively accept this". The first case is done by functions starting with `allow` while the second is done by functions starting with `accept`. I don't know if it's understandable or not…

dcaliste committed:
[voicecall] Add pattern matching for the filter plugin.

Introduce simple patterns matching to
filter incoming calls. New patterns
include prefix matching and all matching.

It also introduces a whitelist in addition
to the ignored and rejected lists.

The priority is always given to exact
number matching before any pattern
matching.

A commhistory helper is used to match
exact numbers ignoring the local or the
international prefix.

dcaliste committed:
[voicecall] Add setters to the filter plugin.

Expand the filter API to add semantic
functions to manipulate ignored and
rejected numbers or range of numbers.

This is packaged in a library to help
downstream packages to handle the filters.

It also adds a test, using this new API,
to ensure stability and conformance of
the filtering plugin.

pvuorela commented on plugins/filter/tests/tests.pro:
```diff
@@ -0,0 +1,22 @@
+TEMPLATE = app
+TARGET = tst_filter
+QT += testlib
+CONFIG -= app_bundle
```
This is maybe some copy-paste origin? I think app_bundle would be valid mostly on macos.

dcaliste replied:
Yes, a copy-paste from nemo-qml-plugin-calendar I think... I didn't know what it was used for. I'll remove this.

pvuorela commented on plugins/filter/src/filter.cpp:
```diff
@@ -59,10 +63,98 @@ QStringList Filter::ignoredList() const
+    ignoreNumber(pattern.prepend('^'));
+}
+
+void Filter::ignoreAll()
```
For blocking all incoming calls in sailfish there's also the call barring option that allows such crude level of blocking.

I'm guessing the idea for this was to ignore all and then explicitly allow something.

dcaliste replied:
Yes, that's the idea. It can be used to do some selective "do not disturb" mode during the night, or a different filtering pattern for a meeting...

pvuorela commented on plugins/filter/src/filter.cpp:
```diff
@@ -59,10 +63,98 @@ QStringList Filter::ignoredList() const
+void Filter::allowAll()
+{
+    d->m_ignored.removeEntry(QStringLiteral("*"));
+    d->m_rejected.removeEntry(QStringLiteral("*"));
```
I could think allowAll() would make the filter allow everything, i.e. clear both of those lists.

If you want to have those ignore/allow by default, maybe these could be called allowByDefault(), ignoreByDefault() etc?

dcaliste replied:
Yes, I like the `ignoreAllByDefault(bool)`, the bool argument being used to add or remove the filter.

dcaliste replied:
Finally, I didn't put a boolean argument but add a `acceptByDefault()` instead that revert the action of `ignoreByDefault()` or `rejectByDefault()`.

pvuorela reviewed (commented):
The first question with these could be how would the api be actually exposed. Now in a plugin it's not usable outside. Either the filter plugin would need to expose some library and header of its own or then something would need to be moved to the main project. If the filter itself is a plugin I think the latter could be too much since it would then define too detailed how filtering plugins should work.

Now the configs are within the same key so I suppose I'd then filter out everything starting with * or ^ on the UI.

Would still be nice to have that literal number N digit matching :) Maybe wouldn't affect this too much, just applied to the exact number match case.

dcaliste commented:
> Would still be nice to have that literal number N digit matching :)

Sorry, I missed your point here : I thought the exact matching is doing this.

> The first question with these could be how would the api be actually exposed.

Yes, it has nothing to do in the plugin. I put it here as a separated commit so we can discuss the API and then we'll see where to put it.

Defining a FilterRule API in voicecall library could be an option. But if the implementation goes in the library, it means that DConf implementation would be enforeced by the library. Which is not my initial idea : the library is providing a filter() API to incoming calls and the library users are free to implement it as they prefer. For example `phonehook` do it differently but can use now the official API to do its job.

I've no clear idea where to put this DConf filtering implementation. Creating a dedicated library, used by the filtering plugin and the telephony component is a bit overkill. I've no good idea for the moment.

dcaliste commented:
I've added three methods `isIgnored(number)`, `isRejected(number)` and `isAccepted(number)` that can be used to label a number as blocked or not in the UI, without having to know the filter rules.

pvuorela commented:
> > Would still be nice to have that literal number N digit matching :)
> Sorry, I missed your point here : I thought the exact matching is doing this.

Now the exact matching case is matching, well, exact numbers. What I'd like to have is same rule as in detecting the contact for incoming number to be used for blocking it. E.g. with Finnish numbers if I  set a contact to have phone number 0501234567, it gets detected as that contact when receiving a call with the international prefix, +358501234567.

Thus, if I block the number via the contact, the call should get blocked too, meaning that the comparison should happen with N last digits based on what we use elsewhere. This shouldn't require declaring any prefix type of rule on the config, but rather happen automatically.

Should there be a similar handling for ^ prefix case? Maybe?  If we don't use that in general UI, I can also live without it, though. Anyhow, if there's a rule declared ^050123 as  a prefix to be blocked, should it then also block the mentioned +358501234567?

Matching phone numbers can be painful.

> I've no clear idea where to put this DConf filtering implementation. Creating a dedicated library, used by the filtering plugin and the telephony component is a bit overkill. I've no good idea for the moment.

Perhaps simplest could be just a separate library created by the plugin here, and installing headers for it in voicecall-qt5-plugin-voicecall-filter-devel (phew) package if we want to keep the literal strings and patterns in same config, and expose the block status to UI at least so much that unblocking automatically creates the whitelist entry if needed.

dcaliste commented:
> Thus, if I block the number via the contact, the call should get blocked too, meaning that the comparison should happen with N last digits based on what we use elsewhere. This shouldn't require declaring any prefix type of rule on the config, but rather happen automatically.

Ok, got it. Yes, blocking via a contact stored with local prefix should match also and is not currently handled. Thanks for the explanations.

> Should there be a similar handling for ^ prefix case? Maybe?

Not sure, it's necessary. To keep it simple, currently prefix handling are entered by hand in the DConf entry. My carrier is always presenting numbers with international prefix, so I've entered dconf values like "^+33XXX". I guess people with a carrier presenting a different behaviour can adjust the values for their own needs. Or even duplicating the values if necessary.

> Perhaps simplest could be just a separate library created by the plugin here

Very nice suggestion, like signon plugins are doing. I'm going to propose something like that. It's consistent, if you use the plugin with the DConf implementation, you can use its own library and headers. Makes sense.

pvuorela commented:
> Not sure, it's necessary. To keep it simple, currently prefix handling are entered by hand in the DConf entry. My carrier is always presenting numbers with international prefix, so I've entered dconf values like "^+33XXX".

Could still note that there is a corner case when roaming abroad. Some operators/countries may use some alternative prefix for incoming calls outside the country. I think we have even some existing problems with those now not always working. Up for consideration how big of a problem that is.

dcaliste commented:
I've librarified the Filter class, so it can be used as a helper elsewhere. I've added the N digit exact matching (and noticed I didn't add a test, will do).

About the carrier while roaming not always presenting the international numbers with the + prefix, well, I don't know how to treat it. One could just ignore it for the moment and handle it later. I don't know how much of a nuisance it can be.

dcaliste commented:
Test added for the last N digit matching. Setting a filter with international prefix will also match incoming numbers without and vice and versa.

dcaliste commented:
Fixed an issue in .pc file generation, thanks to your lastest commit in libcommhistory about missing qmake build parameters ;-)

pvuorela commented on plugins/filter/lib/filter.cpp:
```diff
@@ -0,0 +1,197 @@
+    }
+
+    enum Action {
+          NO_FILTER,
```
Just a suggestion, but how about something like FILTER_APPROVE so all the cases would have similar format? Also this can be now returned if there _is_ a filter, i.e. the white list one.

dcaliste replied:
Nice suggestion, I'll split the NO_FILTER into a FILTER_APPROVE and a FILTER_NO_MATCH. It makes the various possibilities more straight forward. Of course, both APPROVED and NO_MATCH would let the call pass.

pvuorela commented on plugins/filter/lib/lib.pro:
```diff
@@ -0,0 +1,40 @@
+CONFIG += shared \
+          no_install_prl \
+          link_pkgconfig \
+          c++11
```
C++11 already the default. We have a bunch of unnecessary instances of these here and there.

dcaliste replied:
Ok, nice, removed.

pvuorela commented on plugins/filter/lib/voicecall-filter-plugin.pc.in:
```diff
@@ -0,0 +1,11 @@
+
+Name: voicecall-filter-plugin
+Description: Voice call filtering library
+Version: $$PROJECT_VERSION
```
I might just hard-code this to to something to avoid needing to pass the version externally. These can bite back if people start using the project with separate packaging.

dcaliste replied:
Hardcode it requires to remember to update it. Which I often forgot. But I went for the create_pc route. It works fine. Thanks for the tips.

pvuorela commented on plugins/filter/lib/voicecall-filter-plugin.pc.in:
```diff
@@ -0,0 +1,11 @@
+prefix=/usr
+exec_prefix=${prefix}
+libdir=${prefix}/$$PKGCONFIG_LIB
```
Could also use a default for this and avoid need to pass the value externally. Also could consider should qmake be delegated to handle the creation

For reference on both details e.g. https://github.com/sailfishos/nemo-qml-plugin-dbus/blob/master/src/nemo-dbus/nemo-dbus.pro#L53

dcaliste replied:
Very nice that QMake can handle this. I've switch to use it.

pvuorela commented on plugins/filter/lib/lib.pro:
```diff
@@ -0,0 +1,40 @@
+TEMPLATE = lib
+TARGET = voicecall-filter-plugin
```
Since this isn't going under plugins directory, I'd consider it more of a library rather than a plugin. Just voicecall-filter or something?

Likewise if the namespace should be without the plugin part.

dcaliste replied:
That sounds ok. I've removed the plugin mention from the lib (paths, namespace…).

pvuorela commented on plugins/filter/lib/filter.cpp:
```diff
@@ -0,0 +1,197 @@
+    d->m_rejected.removeEntry(QStringLiteral("*"));
+}
+
+void Filter::acceptAll()
```
clearAll() / clear() / clearFilters() or something to highlight it removes everything set?

dcaliste replied:
Well, I don't know if it's a good idea or not, but I tried to name the methods in a way that is independant from the actual three dconf list implementation. After a call to acceptAll(), I don't want to know what happens internally, but all calls are now accepted. If it's implemented with a "*" in the whitelist, leaving all previous filters or by removing the filter is an implementation detail and should not be reflected in the API. All methods should setup the internal lists to do what their names mean, which I think is the case at the moment.

But maybe it's not a good idea… Not sure if on the opposite, the methods should reflect exactly what internally they are doing and let the users create something valid from it. With my approach, I think it's easier to create a UI which can directly call the library methods : want to block a number ? call blockNumber(). want to block a range call blockNumbersStartingWith(), want to unblock a number ? call acceptNumber()…

pvuorela reviewed (commented):
Had a bit of pause here, started looking but running out of time before I have to leave. Posting some comments I managed to write.

dcaliste commented:
Thanks for the review, even if not complete. There are already good improvements from the initial submission. Thanks.

pvuorela commented on plugins/filter/lib/filter.cpp:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    enum Action {
+          FILTER_NO_MATCH,
```
Indentation off here.

dcaliste replied:
Thanks, fixed.

pvuorela commented on plugins/filter/lib/filterlist.cpp:
```diff
@@ -61,7 +64,53 @@ void FilterList::set(const QStringList &list)
+        if (filter.startsWith('+') || filter[0].isDigit()) {
+            // Exact number matching
+            if (number == filter
+                || CommHistory::remoteAddressMatch(CommHistory::RING_ACCOUNT, number, filter, true)) {
```
Uh, this libcommhistory part is icky. First the "commonutils.h" is somewhat dumping ground for random thing, but then this RING_ACCOUNT is just ugly. Moreover the QStringLiteral() there will now create extra instances wherever the header is included.

What if we at least add something like phoneNumberMatch() which would do the same thing but not requiring to pass ring account? And the RING_ACCOUNT we could move away from the public api, I didn't spot any usage outside libcommhistory at the moment. I can also do that.

dcaliste replied:
> What if we at least add something like phoneNumberMatch() which would do the same thing but not requiring to pass ring account? And the RING_ACCOUNT we could move away from the public api, I didn't spot any usage outside libcommhistory at the moment. I can also do that.

I totally agree. If you prefer to concentrate on other tasks, tell me. I can do it also.

pvuorela reviewed (commented):
Didn't try yet, but looks like this should do the trick.

I think I should be able to switch UI side qml components to use this instead of direct dconf. Just need to consider if I want to filter the list of rejected numbers for wildcard, but maybe even that wouldn't be entirely necessary if it's only showing entries set outside the UI.

dcaliste committed:
[voicecall] Use recipient from libcommhistory for number matching.

dcaliste commented:
I've added a commit on top showing how to use Recipient directly to match phone numbers. Now, it is using internally libphonenumber, as Recipient::matchesRemoteUid() is using. It requires https://github.com/sailfishos/libcommhistory/pull/18

I'm wondering if it's the right direction, avoiding to expose more from CommHistory, or if the filterlist::match() method should use directly libphonenumber. It may be simpler, since invoquing IsNumberMatchWithTwoStrings is only a three liner.

pvuorela commented on plugins/filter/src/src.pro:
```diff
@@ -1,16 +1,12 @@
-    filterlist.h
+INCLUDEPATH += $$PWD/../lib
+
+LIBS += -L$$PWD/../lib -lvoicecall-filter-plugin
```
Library name not updated here to skip -plugin. Doesn't compile.

dcaliste replied:
Oops, sorry. Fixed.

pvuorela commented on plugins/filter/tests/tests.xml:
```diff
@@ -0,0 +1,10 @@
+  <suite name="voicecall-filter-plugin" domain="mw">
+    <set name="unit-tests" feature="voicecall-filter-plugin">
+       <case manual="false" name="tst_filter">
+         <step>HOME=$PWD /opt/tests/voicecall/plugins/filter/tst_filter</step>
```
Wrong path.

pvuorela replied:
The HOME part seems to work if not in home directory :)

If you want to avoid using the home, perhaps it could be something hard-coded, e.g. HOME=/opt/tests/voicecall/filter ?

dcaliste replied:
Ah, sure. Corrected. I changed the suite and set names also.

pvuorela commented on plugins/filter/tests/tests.xml:
```diff
@@ -0,0 +1,10 @@
+  <suite name="voicecall-filter" domain="mw">
+    <set name="unit-tests" feature="voicecall-filter">
+       <case manual="false" name="tst_filter">
+         <step>HOME=$PWD /opt/tests/voicecall/filter/tst_filter</step>
```
In case the earlier comment got lost in the update: would you mind hard-coding the HOME to something so this doesn't surprisingly fail when executing from the home directory.

dcaliste replied:
Mmh, is it failing if actually run in the real home directory ? I'm doing this HOME=... not to damage the DConf database, when run on device by the user (me). Maybe the best way would be to hardcode the HOME directory in the test to a temporary directory... I'm going to try this.

pvuorela replied:
Yea, I often just run from the home directory like "testrunner-lite  -f /opt/tests/voicecall/filter/tests.xml -o testrun" and now it started with failure on everything.

Was ok enough if the HOME= was set to directory of the test, don't think it even needs to be a temporary one.

dcaliste replied:
Well, I'm often running the tests like `scp tst_foo nemo@192.168.2.15:~/devel/` and then `cd devel; ./tst_foo`. Having a hardcoded path to something in `/opt/` while I'm not installing the package would make the test fails.

pvuorela replied:
Yea, it wouldn't then work if you want to run the test directly without the xml part. Then again on many places we assume that the xml sets the environment right for the test.

If you get the test running without xml nice, that's nice, but maybe as first step just the HOME in .xml could be adjusted so there wouldn't be this surprise of current working directory affecting the test xml execution?

pvuorela commented:
At least I'm nicely getting now number blocked when configured without an international prefix which makes this good for me by itself already :)

dcaliste commented:
I've just added a temporary folder to be used as the HOME directory. Indeed, when using the real HOME, I put a QVERIFY to check that DConf keys where empty and avoid scratching of user data.

I've removed the HOME=... in the test XML file.

pvuorela reviewed (approved):
Is good, thanks!

lkraav commented on plugins/filter/lib/filter.cpp:
```diff
@@ -0,0 +1,208 @@
+    Private()
+        : m_rejected(QString::fromLatin1("/sailfish/voicecall/filter/rejected-numbers"))
+        , m_ignored(QString::fromLatin1("/sailfish/voicecall/filter/ignored-numbers"))
+        , m_whitelist(QString::fromLatin1("/sailfish/voicecall/filter/whitelist"))
```
Modern equivalent of this term is `allowlist` (+ `blocklist`). Unfortunately didn't know about this repo to review earlier until https://forum.sailfishos.org/t/release-notes-tampella-5-0-0-55/21801/231

Not sure if it's worth it to refactor retroactively, although with initial minimal usage, there will never be a better time.

dcaliste replied:
Hello, thank you for commenting. Do you mean more modern for inclusiveness and avoiding discrimination biais ? Ah, sorry for this I'm a bit too old to easily remember these kind of changes. I've no opinion if it's still early enough to change it or not.

lkraav replied:
> Do you mean more modern for inclusiveness and avoiding discrimination biais?
> Ah, sorry for this I'm a bit too old to easily remember these kind of changes. I've no opinion if it's still early enough to change it or not.

Yes.

I'm also not a fan of spending energy on refactoring old code for such, but this is borderline new and exposes a user-facing API.

It's Jollyboys decision in the end, although a candidate PR could be filed regardless. I think the only semi-tough question is whether to support back-compatibility for those few people who already have entries stored in legacy-named keys :thinking: If on-device keys can be easily renamed, that could simply be a release-based user instruction and problem solved?

pvuorela replied:
I wouldn't care about the modernity per se too much, whitelist has been there for centuries and will be around for some time more. Having that said, "allowed-numbers" type of naming would match the intention and the other configs nicer.

But is it worth adjusting at this point, not sure either. It's a lower level config which isn't directly visible, nor even used, in the UI.
